### PR TITLE
feat(GDB-10638) Add Jenkins pipeline

### PR DESCRIPTION
**Pull request by teodossidossev**
## WHAT:
This commit introduces a Jenkins pipeline for the project, adding automation for build, test, lint, and validation processes across the projects. Along with the Jenkinsfile, several scripts have been added or modified to facilitate the build, testing, and cleaning operations. Additionally, TypeScript and ESLint configurations have been updated to improve code quality and consistency.

## WHY:
To automate the CI/CD process, ensuring that the codebase is consistently built, tested, and validated before being merged.

## HOW:

- Jenkins Pipeline: A new Jenkinsfile is added to define the stages for installing dependencies, linting, validating, building, testing, and running Cypress tests.
- Build Scripts: New shell scripts (install.sh, build.sh, clean.sh, etc.) are introduced for handling installation, building, and cleaning tasks in a modular and reusable manner.
- Linting and Testing: New ESLint configurations and Jest setup for different packages, along with scripts for running linting and tests.
- Configuration Updates: TypeScript configurations are updated to enhance strictness and compatibility, and ESLint configurations are made more consistent across packages.
- Cleanup: Removal of outdated configuration files and cleanup of unnecessary code.

### CODE REVIEW

File: packages/api/src/services/event.service.ts [hunk 1]
```diff
@@ -7,38 +7,38 @@ import {Service} from './service';
...
      *
      * @param event - the event to be emitted.
      * @return the emitted event.
      */
-    emit(event: Event): CustomEvent {
-        const customEvent = new CustomEvent(event.NAME, {detail: event.payload});
-        this.getHostElement().dispatchEvent(customEvent);
-        return customEvent;
-    }
+  emit<T extends {} | undefined>(event: Event<T>): CustomEvent {
```
Comment (svilenvelikov) [inline packages/api/src/services/event.service.ts:16]: Isn't it better the generic types to be defined on the class instead of on each separate method?

File: packages/legacy-workbench/scripts/run-cypress-tests.sh
Comment (svilenvelikov) [inline packages/legacy-workbench/scripts/run-cypress-tests.sh:1]: I've never used this scrip to run tests before. Pavel recently found that it was broken by us somehow and also he introduced some changes here. Are you sure that these changes are correct?

File: packages/shared-components/package.json [hunk 2]
```diff
@@ -36,22 +36,28 @@
     "build:watch": "stencil build --watch",
     "build": "stencil build",
     "start": "stencil build --dev --watch --serve",
-    "test": "stencil test --spec --e2e",
-    "test:watch": "stencil test --spec --e2e --watchAll",
-    "test:spec:watch": "stencil test --spec --watchAll",
-    "test:e2e:watch": "stencil test --e2e --watchAll",
```
Comment (svilenvelikov) [inline packages/shared-components/package.json:42]: these last two commands you deleted are used to run the tests separately, so better restore them

File: packages/shared-components/src/components/onto-header/onto-header.tsx [hunk 3]
```diff
@@ -7,13 +8,32 @@ import { Component, Host, h } from '@stencil/core';
...
+  }
+
   render() {
     return (
       <Host>
         <div class="header-component">
           <div class="search-component">&#x1F50D;</div>
           <div class="repository-selector-component">TestRepo &#8964;</div>
-          <onto-language-selector></onto-language-selector>
+          <div class="language-selector-component" onClick={this.onLanguageChanged}>EN &#8964;</div>
```
Comment (svilenvelikov) [inline packages/shared-components/src/components/onto-header/onto-header.tsx:36]: hmm you need to rebase the branch because the language selector is already in place

File: packages/shared-components/src/components/onto-navbar/onto-navbar.tsx [hunk 4]
```diff
@@ -193,7 +197,7 @@ export class OntoNavbar {
   }
 
   render() {
-    if (!this.menuModel) {
+    if (this.menuModel === null) {
```
Comment (svilenvelikov) [inline packages/shared-components/src/components/onto-navbar/onto-navbar.tsx:200]: why ? :^/

File: packages/shared-components/src/components/onto-navbar/onto-navbar.tsx [hunk 5]
```diff
@@ -223,7 +227,7 @@ export class OntoNavbar {
               {item.children.length > 0 &&
                 <Fragment>
                   <div class={{'menu-element-root': true, 'active': item.selected}}
-                       onClick={(event) => this.select(event, item)}>
+                       onClick={this.handleSelectMenuItem(item)}>
```
Comment (svilenvelikov) [inline packages/shared-components/src/components/onto-navbar/onto-navbar.tsx:230]: we need to discuss these

File: packages/shared-components/src/components/translate-label/translate-label.tsx [hunk 6]
```diff
@@ -21,7 +21,16 @@ export class TranslateLabel {
 
   private readonly unsubscribeTranslationChanged: Function;
 
+  /**
+   * Represents a label key.
+   * @typedef {string} labelKey
```
Comment (svilenvelikov) [inline packages/shared-components/src/components/translate-label/translate-label.tsx:26]: I think the typedef's are redundant here. Remove them

File: packages/shared-components/stencil.config.ts [hunk 7]
```diff
@@ -8,17 +8,12 @@ export const config: Config = {
       type: 'dist',
       esmLoaderPath: '../loader',
     },
-    // {
-    //   type: 'dist-custom-elements',
-    //   customElementsExportBehavior: 'single-export-module',
-    //   externalRuntime: false,
-    // },
```
Comment (svilenvelikov) [inline packages/shared-components/stencil.config.ts:15]: Why did you remove this?

File: packages/workbench/sonar-project.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+const sonarqubeScanner = require("sonarqube-scanner");
+
+const branch = argv['branch'];
+const targetBranch = argv['target-branch'];
+const pullRequestId = argv['pull-request-id'];
+
+const sonarOptions = {
+  "sonar.organization": "ontotext-ad",
+  "sonar.projectKey": "Ontotext-AD_WB_migration_workbench",
+  "sonar.projectName": "WB Migration: workbench",
```
Comment (svilenvelikov) [inline packages/workbench/sonar-project.js:11]: Can we change the key and the name please! And same for the shared components where you also added the `WB Migration ` prefix

File: scripts/lint.sh [hunk 9]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+
+# This script should be run from the root directory of the project.
+
+# Source the error handling script
+. scripts/error-handling.sh
+
+install_package() {
```
Comment (svilenvelikov) [inline scripts/lint.sh:8]: this should be lint_package

File: packages/shared-components/src/components/onto-dropdown/onto-dropdown.tsx [hunk 10]
```diff
@@ -84,35 +85,43 @@ export class OntoDropdown {
...
                             <span class={'onto-dropdown-option-icon ' + item.iconClass}></span>
-                            <span innerHTML={item.name ? item.name : TranslationService.translate(item.nameLabelKey)}></span>
+                            <span innerHTML={item.name ?? TranslationService.translate(item.nameLabelKey)}></span>
                         </button>)}
                 </div>
             </div>
         );
     }
 
+    private toggleComponentHandle() {
```
Comment (svilenvelikov) [inline packages/shared-components/src/components/onto-dropdown/onto-dropdown.tsx:117]: typo

File: packages/shared-components/src/components/onto-language-selector/onto-language-selector.tsx [hunk 11]
```diff
@@ -54,6 +55,10 @@ export class OntoLanguageSelector {
     this.items = this.getLanguageDropdownOptions();
   }
 
+  private onLanguageChangedHandle() {
```
Comment (svilenvelikov) [inline packages/shared-components/src/components/onto-language-selector/onto-language-selector.tsx:58]: typo

File: packages/api/src/services/event.service.ts (hunk 1 above)
Comment (teodossidossev) [reply packages/api/src/services/event.service.ts:16]: it's more flexible on the methods

File: packages/legacy-workbench/scripts/run-cypress-tests.sh
Comment (teodossidossev) [reply packages/legacy-workbench/scripts/run-cypress-tests.sh:1]: those are Pavel's changes

File: packages/shared-components/package.json (hunk 2 above)
Comment (teodossidossev) [reply packages/shared-components/package.json:42]: done

File: packages/shared-components/src/components/onto-navbar/onto-navbar.tsx (hunk 5 above)
Comment (teodossidossev) [reply packages/shared-components/src/components/onto-navbar/onto-navbar.tsx:234]: we did

File: packages/shared-components/src/components/translate-label/translate-label.tsx (hunk 6 above)
Comment (teodossidossev) [reply packages/shared-components/src/components/translate-label/translate-label.tsx]: done

File: packages/shared-components/stencil.config.ts (hunk 7 above)
Comment (teodossidossev) [reply packages/shared-components/stencil.config.ts]: reverted

File: scripts/lint.sh (hunk 9 above)
Comment (teodossidossev) [reply scripts/lint.sh]: renamed

File: scripts/sonar.sh [hunk 12]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+
+# This script should be run from the root directory of the project.
+
+# Source the error handling script
+. scripts/error-handling.sh
+
+install_package() {
```
[the remark this answers is not in the document]
Comment (teodossidossev) [reply scripts/sonar.sh]: yup

File: packages/shared-components/src/components/onto-dropdown/onto-dropdown.tsx (hunk 10 above)
Comment (teodossidossev) [reply packages/shared-components/src/components/onto-dropdown/onto-dropdown.tsx]: fixed

File: packages/workbench/sonar-project.js (hunk 8 above)
Comment (teodossidossev) [reply packages/workbench/sonar-project.js]: done

File: packages/shared-components/src/components/onto-navbar/onto-navbar.tsx (hunk 4 above)
Comment (teodossidossev) [reply packages/shared-components/src/components/onto-navbar/onto-navbar.tsx]: fixed

File: packages/api/src/services/authentication.service.ts [hunk 13]
```diff
@@ -2,6 +2,6 @@ import {Service} from './service';
 
 export class AuthenticationService implements Service {
   login(): string {
-    return 'Authentication.login from the API';
+    return 'Athentication.login from the API';
```
Comment (svilenvelikov) [inline packages/api/src/services/authentication.service.ts:5]: interesting

Comment (teodossidossev) [reply packages/api/src/services/authentication.service.ts]: fixed

**Review by svilenvelikov — APPROVED** [review]
